File: src/deepgo/native/cpp/Processor.cpp
#include "Processor.h"

#include <algorithm>
#include <iostream>

namespace deepgo {

/**
 * Creates a computation management object.
 * @param model Model file
 * @param gpus List of GPU numbers
 * @param batchSize Maximum batch size
 * @param fp16 True to compute with 16-bit precision
 * @param deterministic True to make computation results reproducible
 * @param threadsParGpu Number of threads per GPU
 */
Processor::Processor(
    std::string model, std::vector<int32_t> gpus, int32_t batchSize,
    bool fp16, bool deterministic, int32_t threadsParGpu)
    : _mutex(),
      _executors() {
  for (auto gpu : gpus) {
    for (int32_t i = 0; i < threadsParGpu; i++) {
      _executors.push_back(
          std::make_unique<Executor>(model, gpu, batchSize, fp16, deterministic));
    }
  }
}

/**
 * Executes inference.
 * @param inputs Input data
 * @param outputs Output data
 * @param size Number of input/output data
 */
void Processor::execute(float* inputs, float* outputs, int32_t size) {
  // Select the computation executor to run next
  int32_t min_index = 0;

  {  // synchronize
    // Find the computation executor with the fewest waiting tasks
    std::unique_lock<std::mutex> lock(_mutex);
    int32_t min_count = _executors[0]->getWaitingCount();

    for (int32_t i = 1; i < _executors.size(); i++) {
      int32_t count = _executors[i]->getWaitingCount();

      if (count < min_count) {
        min_index = i;
        min_count = count;
      }
    }

    // Increase the reserved count
    _executors[min_index]->addReservedCount(size);
  }

  // Execute the computation
  _executors[min_index]->execute(inputs, outputs, size);
}

}  // namespace deepgo

File: src/deepgo/processor.py
from pathlib import Path
from typing import List

import numpy as np

from .config import DEFAULT_BATCH_SIZE, DEFAULT_THREADS_PER_GPU
from .native import NativeInferenceProcessor


class Processor(object):
    def __init__(
        self,
        model: str | Path,
        gpus: List[int] = [-1],
        batch_size: int = DEFAULT_BATCH_SIZE,
        fp16: bool = False,
        deterministic: bool = False,
        threads_per_gpu: int = DEFAULT_THREADS_PER_GPU,
        cache_size: int = 0,
    ) -> None:
        '''Initialize computation management object.
        Args:
            model (str | Path): Path to model file
            gpus (List[int]): List of GPU numbers to use
            batch_size (int): Maximum batch size
            fp16 (bool): True to use FP16
            deterministic (bool): True to make results reproducible
            threads_per_gpu (int): Number of threads per GPU
            cache_size (int): Cache size for board evaluation
        '''
        if not Path(model).exists():
            raise FileNotFoundError(f'File not found: {model}')

        self.native = NativeInferenceProcessor(
            str(model), gpus, fp16, deterministic, batch_size, threads_per_gpu, cache_size)

    def execute(self, inputs: np.ndarray) -> np.ndarray:
        '''Execute inference.
        Args:
            inputs (np.ndarray): Input data
        '''
        return self.native.execute(inputs)

    def get_batch_fill_rate(self) -> float:
        '''Get the ratio of inference requests included in the batch.
        Returns:
            float: Ratio of inference requests included in the batch
        '''
        return self.native.get_batch_fill_rate()

    def get_cache_hit_rate(self) -> float:
        '''Get the cache hit rate of inference.
        Returns:
            float: Cache hit rate of inference
        '''
        return self.native.get_cache_hit_rate()
